// remarkable-pgn/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _pgn4web = require('./pgn4web');

var getLine = function getLine(state, line) {
  var pos = state.bMarks[line] + state.blkIndent,
      max = state.eMarks[line];

  return state.src.substr(pos, max - pos);
};

var pgnRule = function pgnRule(state, startLine, endLine, checkMode) {
  if (getLine(state, startLine) == '[PGN]') {
    var startPgn = startLine + 1;
    var nextLine = startPgn;
    while (nextLine < endLine) {
      if (getLine(state, nextLine) == '[/PGN]') {
        state.tokens.push({
          type: 'pgn',
          content: state.getLines(startPgn, nextLine, state.blkIndent, true),
          block: true,
          lines: [startLine, nextLine],
          level: state.level
        });
        state.line = nextLine + 1;
        return true;
      }
      nextLine++;
    }
  }
  return false;
};

var render_pgn4web = function render_pgn4web(tokens, idx, options) {
  var _generate = (0, _pgn4web.generate)(tokens[idx].content, options.pgn),
      url = _generate.url,
      frameHeight = _generate.frameHeight,
      frameWidth = _generate.frameWidth;

  return '<iframe\n      height="' + frameHeight + '" width="' + frameWidth + '" src="' + url + '"\n      scrolling=\'no\' frameBorder=\'0\'\n      marginHeight=\'0\' marginWidth=\'0\'>\n    Your web browser do not support iframes as required to display the chessboard\n  </iframe>';
};

exports.default = function (md, options) {
  md.set({ pgn: options });
  md.block.ruler.before('code', 'pgn', pgnRule, {});
  md.renderer.rules.pgn = render_pgn4web;
};
